### PR TITLE
Increment 9

### PR DESCRIPTION
**Pull request by Taboon**

### CODE REVIEW

File: cmd/shortener/main.go [hunk 1]
```diff
@@ -2,19 +2,40 @@ package main
...
+
+	if err := logger.Initialize(conf.LogLevel); err != nil {
+		log.Fatal("Can't set logger")
+	}
+
+	var stor storage.Repository
+
+	if conf.FileBase.File != "" {
+		logger.Log.Info("Используем файловое хранилище", zap.String("file", conf.FileBase.File))
+		stor = storage.NewFileStorage(conf.FileBase.File)
```
Comment (arttet) [inline cmd/shortener/main.go]: CHANGE REQUEST: it shouldn't be a file system storage. It looks like a backup.

File: internal/config/config.go [hunk 2]
```diff
@@ -31,34 +50,37 @@ func (l *Address) String() string {
 // и устанавливает правила парсинга для пользовательского типа.
 func (l *Address) Set(flagValue string) error {
 
+	if flagValue == "" {
+		return errors.New("пустое значение флага")
```
Comment (arttet) [inline internal/config/config.go]: CHANGE REQUEST: please create a variable for it.

File: internal/config/config.go [hunk 3]
```diff
@@ -31,34 +50,37 @@ func (l *Address) String() string {
...
-		return err
-	}
-	l.IP = address[0]
 
-	if address[1] == "" {
-		err := errors.New("wrong port")
-		return err
-	}
-	port, err := strconv.Atoi(address[1])
+	for i, v := range address {
```
Comment (arttet) [inline internal/config/config.go]: CHANGE REQUEST: too complicated

File: internal/config/config.go [hunk 4]
```diff
@@ -70,46 +92,48 @@ func (c *Config) BuildConfig() Config {
 		},
 	}
 
-	err := c.parseEnv(&conf)
+	conf.FileBase.File = "/tmp/short-url-db.json"
```
Comment (arttet) [inline internal/config/config.go]: CHANGE REQUEST: this is a magic string

File: internal/logger/logger.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+	return size, err
+}
+
+func (r *loggingResponseWriter) WriteHeader(statusCode int) {
+	// записываем код статуса, используя оригинальный http.ResponseWriter
+	r.ResponseWriter.WriteHeader(statusCode)
+	r.responseData.status = statusCode // захватываем код статуса
+}
+
+var Log *zap.Logger = zap.NewNop()
```
Comment (arttet) [inline internal/logger/logger.go]: CHANGE REQUEST: please don't use global variables

File: internal/server/handlers.go [hunk 6]
```diff
@@ -49,7 +60,52 @@ func (s *Server) getURL(w http.ResponseWriter, r *http.Request) {
...
+}
+
+func (s *Server) shortenJSON(w http.ResponseWriter, r *http.Request) {
+	req, err := io.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, "Не удалось прочитать запрос", http.StatusBadRequest)
+		return
+	}
+
+	body := Request{}
```
Comment (arttet) [inline internal/server/handlers.go]: CHANGE REQUEST: this is a magic name.

File: internal/server/handlers.go [hunk 7]
```diff
@@ -49,7 +60,52 @@ func (s *Server) getURL(w http.ResponseWriter, r *http.Request) {
...
+	fmt.Println(string(req))
+	err = json.Unmarshal(req, &body)
```
Comment (arttet) [inline internal/server/handlers.go]: CHANGE REQUEST: it would be better if it is in the `validate` package

File: internal/server/handlers.go [hunk 8]
```diff
@@ -49,7 +60,52 @@ func (s *Server) getURL(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+
+	id, err := s.urlSaver(url)
+	if err != nil {
+		http.Error(w, "Не удалось сохранить URL: "+err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	response := Response{Result: httpPrefix + s.Conf.BaseURL.String() + "/" + id}
```
Comment (arttet) [inline internal/server/handlers.go]: CHANGE REQUEST: please use formatting strings

File: internal/storage/file_storage.go [hunk 9]
```diff
@@ -0,0 +1,169 @@
+package storage
+
+import (
+	"bufio"
+	"encoding/json"
+	"errors"
+	"github.com/Taboon/urlshortner/internal/logger"
```
Comment (arttet) [inline internal/storage/file_storage.go]: CHANGE REQUEST: please sort imports

File: internal/storage/internal.go [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+type SafeMap struct {
+	mu             sync.Mutex
+	mapStor        map[string]string
+	reverseMapStor map[string]string
+}
+
+var _ Repository = (*SafeMap)(nil)
+
+func NewInternalStorage() *SafeMap {
```
Comment (arttet) [inline internal/storage/internal.go]:
```suggestion
func NewMemoryStorage() *SafeMap {
```

File: test.txt
Comment (arttet) [inline test.txt:1]: CHANGE REQUEST: please remove it

**Review by arttet — CHANGES_REQUESTED** [review]

File: internal/config/config.go (hunk 2 above)
Comment (Taboon) [reply internal/config/config.go]: Done

File: internal/config/config.go (hunk 4 above)
Comment (Taboon) [reply internal/config/config.go:97]: move to const

File: internal/config/config.go (hunk 2 above)
Comment (Taboon) [reply internal/config/config.go]: done

File: internal/config/config.go (hunk 4 above)
Comment (Taboon) [reply internal/config/config.go]: moved to const

File: internal/logger/logger.go (hunk 5 above)
Comment (Taboon) [reply internal/logger/logger.go:35]: Quote from 3 lessons - 2 sprints:
// Log будет доступен всему коду как синглтон.
// Никакой код навыка, кроме функции InitLogger, не должен модифицировать эту переменную.
// По умолчанию установлен no-op-логер, который не выводит никаких сообщений.
var Log *zap.Logger = zap.NewNop()

I'm confused. Is there incorrect information in the lesson?

File: internal/server/handlers.go (hunk 7 above)
Comment (Taboon) [reply internal/server/handlers.go]: done

File: internal/server/handlers.go (hunk 8 above)
Comment (Taboon) [reply internal/server/handlers.go]: Done

File: internal/storage/file_storage.go (hunk 9 above)
Comment (Taboon) [reply internal/storage/file_storage.go]: done

File: internal/storage/internal.go (hunk 10 above)
Comment (Taboon) [reply internal/storage/internal.go]: done

File: test.txt
Comment (Taboon) [reply test.txt:1]: done

File: internal/config/config.go (hunk 3 above)
Comment (Taboon) [reply internal/config/config.go]: done

Comment (Taboon) [reply internal/config/config.go]: simplyfied

File: internal/server/handlers.go (hunk 6 above)
Comment (Taboon) [reply internal/server/handlers.go]: done

File: cmd/shortener/main.go (hunk 1 above)
Comment (Taboon) [reply cmd/shortener/main.go]: done

File: internal/logger/logger.go (hunk 5 above)
Comment (Taboon) [reply internal/logger/logger.go]: done

File: internal/config/config.go [hunk 11]
```diff
@@ -31,34 +57,32 @@ func (l *Address) String() string {
...
-	l.Port = port
 	return nil
 }
 
 func (c *Config) URL() string {
 	return fmt.Sprintf("%v:%v", c.LocalAddress.IP, strconv.Itoa(c.LocalAddress.Port))
 }
 
-func (c *Config) BuildConfig() Config {
+func BuildConfig() *Config {
```
Comment (arttet) [inline internal/config/config.go]: SUGGESTION: https://refactoring.guru/design-patterns/builder/go/example

File: internal/domain/usecase/getURL.go [hunk 12]
```diff
@@ -0,0 +1,20 @@
+package usecase
+
+import (
+	"errors"
+
+	"github.com/Taboon/urlshortner/internal/storage"
+)
+
+var errUnknownID = errors.New("отсутствует такой ID")
```
Comment (arttet) [inline internal/domain/usecase/getURL.go]: CHANGE REQUEST: it should be in another package such as `error`

Comment (arttet) [inline internal/domain/usecase/getURL.go]: CHANGE REQUEST: please use English

File: internal/logger/logger.go [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+	return size, err
+}
+
+func (r *loggingResponseWriter) WriteHeader(statusCode int) {
+	// записываем код статуса, используя оригинальный http.ResponseWriter
+	r.ResponseWriter.WriteHeader(statusCode)
+	r.responseData.status = statusCode // захватываем код статуса
+}
+
+var log = &zap.Logger{}
```
Comment (arttet) [inline internal/logger/logger.go]: CHANGE REQUEST: please don't use global variables

File: internal/server/handlers.go [hunk 14]
```diff
@@ -32,24 +43,73 @@ func (s *Server) getURL(w http.ResponseWriter, r *http.Request) {
...
+	_, err = w.Write([]byte(fmt.Sprintf("%s%s/%s", httpPrefix, s.Conf.BaseURL.String(), id)))
+
+	if err != nil {
+		http.Error(w, "Не удалось записать ответ: "+err.Error(), http.StatusBadRequest)
+		return
+	}
+}
+
+var requestBody = &RequestJSON{}
+var response = &Response{}
```
Comment (arttet) [inline internal/server/handlers.go]: CHANGE REQUEST: please find bugs here

File: internal/storage/file_storage.go [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+		}
+	}()
+
+	if err != nil {
+		return URLData{}, false, err
+	}
+	scanner := bufio.NewScanner(file)
+	var shortUrls []URLData
+
+	for scanner.Scan() {
```
Comment (arttet) [inline internal/storage/file_storage.go:72]: CHANGE REQUEST: please provide me asymptotic estimation about this method.

**Review by arttet — CHANGES_REQUESTED** [review]

File: internal/logger/logger.go (hunk 13 above)
Comment (Taboon) [reply internal/logger/logger.go]: Fix

File: internal/domain/usecase/getURL.go (hunk 12 above)
Comment (Taboon) [reply internal/domain/usecase/getURL.go]: done

Comment (Taboon) [reply internal/domain/usecase/getURL.go]: done

File: internal/config/config.go (hunk 11 above)
Comment (Taboon) [reply internal/config/config.go]: Accepted

File: internal/storage/file_storage.go (hunk 15 above)
Comment (Taboon) [reply internal/storage/file_storage.go:72]: Optimised

File: internal/server/handlers.go (hunk 14 above)
Comment (Taboon) [reply internal/server/handlers.go]: Non-thread-safe code
Fixed

File: cmd/shortener/main.go [hunk 16]
```diff
@@ -1,20 +1,62 @@
...
+	"log"
+
 	"github.com/Taboon/urlshortner/internal/config"
+	"github.com/Taboon/urlshortner/internal/domain/usecase"
 	"github.com/Taboon/urlshortner/internal/server"
 	"github.com/Taboon/urlshortner/internal/storage"
-	"log"
 )
 
+const baseFilePath = "/tmp/short-url-db.json"
```
Comment (arttet) [inline cmd/shortener/main.go:13]: This is a full path

**Review by arttet — APPROVED** [review]

**Pull request merged**